File: src/year2015/day01/part2.cc
#include <cstdio>
#include <cstdlib>
#include <stdexcept>
#include <string>

#include "../solutions.h"

auto Year2015::Day01::Part2(const std::string& input) -> std::string {
  int floor = 0;
  for (size_t i = 0; i < input.size(); i++) {
    switch (input[i]) {
      case '(':
        floor++;
        break;
      case ')':
        floor--;
        break;
      default:
        throw std::runtime_error(
            "Invalid character encountered while reading input.txt");
    }
    if (floor == -1) {
      return std::to_string(i + 1);
    }
  }
  throw std::runtime_error("Santa never reached the basement");
}

File: src/year2015/day01/part2.rs
pub fn part2(input: String) -> Result<String, String> {
  let mut floor = 0;
  for i in 0..input.len() {
    match input.bytes().nth(i).unwrap() {
      b'(' => floor += 1,
      b')' => floor -= 1,
      _ => {
        return Err("Invalid character encountered while reading input.txt".to_string());
      }
    }
    if floor == -1 {
      return Ok((i + 1).to_string());
    }
  }
  return Err("Santa never reached the basement".to_string());
}
